### PR TITLE
Use Huffman compression for blackbox data transfer

### PR DESCRIPTION
**Pull request by martinbudden**
This includes an implementation of Huffman using a pre-computed Huffman table.

See issue https://github.com/betaflight/betaflight/issues/1156 - ""Suggestion: Use compression to speed up on-board flash blackbox transfer for discussion" and PR https://github.com/betaflight/betaflight/issues/1198

Initial implementation in Cleanflight because it has a working test harness.

### CODE REVIEW

**Comment by etracer65** [comment]
Since I'm expecting the "what is the optimal tree" conversation to continue…  I'm wondering how feasible it would be to always calculate the optimal tree for the data rather than using a predefined table?  Does anybody know the read speed from SPI flash?  In a worst case we'd have to read 16MB (128mbit flash chip is the biggest I know of)  and calculate the frequencies (256 entry table).  Then generating the tree is relatively simple and not processing intensive.  If it takes a few seconds to read through the flash during the analysis stage than it's probably worth it.  Significantly longer than probably not.

**Comment by martinbudden** [comment]
I've done some analysis of the BB log files and there format is such that I expect there to be little to be gained by generating the optimal tree.

If you get 30% compression with an optimal tree and 29% compression with a predefined tree it's just not worth it. Even if the numbers are 30% and 25% it's still not worth it.

Anyway, once this is working the facility to generate an optimal tree can be added later. I'm not going to do any work on that myself though.

**Comment by ledvinap** [comment]
Any real-word performance data?

**Comment by martinbudden** [comment]
Yes, I've done some cli based Huffman compression of blackbox logs and get about 30% compression.

**Comment by ledvinap** [comment]
@martinbudden : Its probably easier to implement test code as separate utility...

If the coding is augmented to handle consecutive 0xff/0x00, it may be possible to drop some coding optimizations from blackbox (variable length encoding, maybe others) and get about some output bitrate.

BTW: gzip/bzip2 packs files to 78% of original ... your results are surprisingly good ..

File: src/main/common/huffman.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+int huffmanEncodeBuf(uint8_t *outBuf, int outBufLen, const uint8_t *inBuf, int inLen, const huffmanTable_t *huffmanTable)
+{
+    int ret = 0;
+
+    uint8_t *outByte = outBuf;
+    *outByte = 0;
+    uint8_t outBit = 0x80;
+    for (int ii = 0; ii< inLen; ++ii) {
```
Comment (ledvinap) [inline src/main/common/huffman.c]: this code can be greatly optimized, especially on ARM ...

File: src/main/common/huffman.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+#pragma once
+
+#include <stdint.h>
+
+#define HUFFMAN_EOF (-1)
+
+#define HUFFMAN_TABLE_SIZE 257 // 256 characters plus EOF
+typedef struct huffmanTable_s {
```
Comment (ledvinap) [inline src/main/common/huffman.h]: This table uses 32 bits per entry, 16 bits should be enough (12 for code, 4 for code length)

File: src/main/common/huffman.c (hunk 1 above)
Comment (martinbudden) [inline src/main/common/huffman.c]: I follow the "get it working" and then "get it working well" approach. So optimisation can come later.

But if you want to make a PR I'll be happy to take it.

File: src/main/common/huffman.h (hunk 2 above)
Comment (martinbudden) [inline src/main/common/huffman.h]: Yes, I thought about it. The current table only uses 12 bits for code. However I want to run the table generator over other blackbox logs before I freeze the Huffman table. Other input files may generate a Huffman table that uses 13 or more bits. So I've kept the flexibility.

**Comment by martinbudden** [comment]
@ledvinap , the cli Huffman code and the tree generation code is now at:

https://github.com/martinbudden/huffman/tree/tree_output

File: src/main/common/huffman.h (hunk 2 above)
Comment (ledvinap) [inline src/main/common/huffman.h]: IMO to get code longer than 12 bit you need symbol with relative frequency lower than 2^(-12) (or similar). With 8bit table this should not be a problem

Comment (martinbudden) [inline src/main/common/huffman.h]: Well, I've only tried a few black box logs, but I've already seen ratios of 1200:1 (there are an awful lot of zeros in BB logs). So I wouldn't discount the possibility of ratios in excess of 4096:1. I need to get hold of a variety of logs from different sources.

**Comment by ledvinap** [comment]
@martinbudden : Can you send me some BB logs? 
Ratio between 0x00 and least used char is ~ 159 in .LOGs I have here

```
perl -ne '
  ++$c{$_} foreach split //;
  END { printf "0x%02x : %d\n", ord($_), $c{$_}
        foreach (sort {$c{$b} <=> $c{$a}} keys %c) }' LOG*.TXT

**Comment by martinbudden** [comment]
@ledvinap , I've attached a couple of black box logs. Note in the second one, 13 bit codes are used. It's a bit of an odd one though - the code for zero is 0 - a one-bit code. That file must be full of zeros.

The logs I am using may not be entirely representative - can you send me a couple of your logs so I have a wider data set to work with?

**Comment by etracer65** [comment]
@martinbudden Here are three logs from this past weekend to add to your analysis.  Three different quads, but all running BF 3.0 and using the SPRACINGF3 target (X-Racer F303's).  Logging 1:1 at 2.67KHz with debug_mode=NOTCH.

[etracer65_blackbox_logs.zip](https://github.com/cleanflight/cleanflight/files/485875/etracer65_blackbox_logs.zip)

**Comment by martinbudden** [comment]
@etracer65 , thanks for the logs.

I just checked the first log file - got 42% compression. Ratio of most frequent character (0x00) to least frequent character (0xFA) was 775394:298 or about 2602:1. `0x00` encoded to `11` and `0xFA` encoded to `00000000000001` - a 14-bit code.

**Comment by ledvinap** [comment]
Interesting.. 
Longest code length affects worst case performance, so it is desirable to limit it, at least for fixed table.

I am very surprised by Huffman coding performance.. I would expect that coding tricks in BB will remove a lot more of the entropy..

**Comment by etracer65** [comment]
@martinbudden 

> @etracer65 , thanks for the logs.

> I just checked the first log file - got 42% compression. Ratio of most frequent character (0x00) to least frequent character (0xFA) was 775394:298 or about 2602:1. 0x00 encoded to 11 and 0xFA encoded to 00000000000001 - a 14-bit code.

If you get a chance, try testing the compression on this same file using the encoding table from your code.  Since this file generates a significantly different table it would be interesting to see how much is being "given up" by using a standardized table vs. an optimal code table.  If it still comes in ~30% then I don't think there's any concern.  Yes, some is being left on the table but 30% is still good.  I'd be concerned if the non-optimal table significantly lowers the results.

I suspect the high logging rate and notch filter debugging is probably what's making this file different.  But it's good to see different examples.

**Comment by ledvinap** [comment]
Some logs I collected are here:
https://drive.google.com/file/d/0B9HM4UOKDQBLTzVfVDM0Q3EzX1k/view?usp=sharing
It's mix of logs from different users, some may be experimental/damaged

**Comment by ledvinap** [comment]
All my log files together compress to 78.6% - 21% saved

**Comment by hydra** [comment]
Hi guys.

While this is all very interesting I think it'd be better if the IO code was done before we look at this.  Right now I'd rather get to a point were it's easier for the 3 forks to share code before adding more code.

**Comment by martinbudden** [comment]
@hydra , this PR has a while to go before it is ready to merge - it needs corresponding configurator changes for a start.

As for the new IO, I agree that is very important. I'm waiting for 1.14 to come out before I do any work on that, since it involves disruptive changes.

**Comment by martinbudden** [comment]
@ledvinap , yes your logs do compress less. They have fewer zeroes in them. I'm not sure why - perhaps the other logs collected data when the quad was stationary on the ground?

Even so, we are getting compression of 20% to 40% depending on log file, which I think is a useful saving.

@etracer65 , some figures from your logs, which I'll call b1, b2 and b3.

logfile | size | compressed size | saving | compressed with b1 Huffman table | saving
------------ | ------------- |------------ | -------------  | ------------- | ---
b1 | 2160640 | 1267940 | 41.3 | 1267683* | 40.8
b2 | 2408448 | 1648839| 31.6 | 1648582 | 31.4
b3 | 2336768 | 1565289 | 33.0| 1567547 | 32.9

* For b1 the second compression is done with the Huffman table from by origin test file `bblog.txt`

So, for these log files at least, we get good compression with a common Huffman tree. I don't find this surprising because a large proportion of the compression was because of the large numbers of certain byte values, in particular: 0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x50

I expect the results will not be so favourable with @ledvinap 's log files - I will investigate this.

**Comment by ledvinap** [comment]
@martinbudden : Most of my log files are recorded at around 3ms looptime. Also some of my logs include unfiltered gyro at 1kHz (using multiple fields), lot of entropy there.

**Comment by ledvinap** [comment]
Huffman disadvantage is that symbol frequency has to be rounded to power of two. With very high frequency of some symbols, this degrades performance considerably.
Arithmetic coding is much better in this regard and it may be feasible : https://github.com/rygorous/ryg_rans/
I'am getting results that are only slightly worse than bzip2 and much better than gzip ... 
Encoding table can be quite small (32bits per symbol) and it may take reasonable number of cycles (even when assuming 12 cycle divide)

**Comment by ledvinap** [comment]
https://github.com/rygorous/ryg_rans/blob/master/rans_word_sse41.h, see encoder

per symbol: 1x divide, few shifts by constant and sums

output in 16bit units, so it will avoid bitstuffing from Huffman (no need for optimization/output buffer in register)

Results are great, but it may be caused by distribution based on actual data, fixed table may make it worse.
Maybe per-quad table will bring considerable improvement for low-noise quads ...

**Comment by hydra** [comment]
@martinbudden ok.

**Comment by martinbudden** [comment]
@ledvinap , let's not the best be the enemy of the good here. Huffman compression is working and gives useful results. There are better compression methods, but it it will take some time to implement any of them.

I am planning to add a byte to the `MSP_DATAFLASH_READ` message which specifies compression type (or no compression), so it will be possible to add further compression methods in future.

Indeed on faster processors it will be possible, on a message per message basis, try several compression methods and use the compression method that is best for that message.

Another simple way to improve compression would to change the format of the interframe - P frames. These are so common that `P` is about the 6th most common character in log files. If interframes could be delimited with a character that is already common, say, `0x01` then better compression could be achieved. I don't know enough about how the black box logs are interpreted to know if that would be possible.  (`0x00 would be even better, but from reading the code that looks like it would cause conflicts)

**Comment by ledvinap** [comment]
@martinbudden :  From my perspective transfer compression is useful only as proof of concept.
Much more interesting is 
- compression of logs before storage 
 - effectively increasing possible transfer rate (but interesting parts may be harder to compress, possibly dropping these parts)
 - increasing onboard flash capacity 

- replacing byte stuffing tricks/optimizations in encoder with entropy coder.
 - Not sure if this will work ... but it may greatly simplify blackbox encoding, making it easier to extend it ..

**Comment by etracer65** [comment]
@ledvinap Transfer compression is much more useful than simply as a proof of concept.  Spending ~30% less time downloading dataflash logs will have a meaningful impact.  Anybody who has ever tried to tune in the field will agree.

I agree with @martinbudden that this is a good first step.  It demonstrates the usefulness of compression in this case and provides a real benefit.  Initially compression was rejected because it was assumed that since the logs were binary they wouldn't be compressible.  In reality they turned out to be quite low entropy and compression has proven to be viable.

There may be opportunities to extend this to compressing the logs before being written to the log device as suggested.  From a dataflash perspective the only real gain there would be additional capacity.  Transfer speed hasn't (yet?) proven to be a constraint.  SAme for the (relatively) few boards that have SPI-connected SD card slots.  For serial openlog devices the reduction in bandwidth is most important.  In the big picture though I'm not sure how much benefit the extra 30% will help on the already quite slow serial rate.  And as stated, compression isn't always linear or predictable.  There may be high entropy periods in the log that result in the compression actually growing the data.  This would be terrible from a serial openlog perspective.  And I have to be realistic and say that serial logging is rapidly dying out.  The PID loops are just getting too fast and there's too much additional data going into the logs requiring large logging denominators - making the logs not that useful because of the low resolution data that remains.

Lastly there's the issue of processing power and whether we want to incur the additional load during the PID loop as we write data to storage.  Would likely be a real problem on F1 targets, probably on F3 - at least impacting max gyro/PID loop capabilities.  Using compression during download has none of these concerns.

**Comment by ledvinap** [comment]
@etracer65 : Yes, I mostly agree with you points.

Openlog speed was mainly limited by SD page erase latency (and small, ~1k buffer in openlog), not serial speed. It's possible to run blackbox port at 500kbit/s (tested), HW limit is (guessing) 4.5 or 9Mbit/s

@thenickdude probably improved this on onboard SD cards by preallocating all blocks, same trick may work for openlog

Serial logging to Intel Edison is quite nice - automatic log upload over Wifi, wireless (WIFI) MSP, lot of onboard storage .. 

And processing power - both Huffman and rANS may be optimized to performance comparable/better than current serial port write overhead. If blackbox coding can be simplified (variable length integer packing is quite expensive), it may result in overall gain in performance

**Comment by digitalentity** [comment]
Interesting. If we log blackbox data without variable-length encoding
(frames of fixed size) and simply feed the byte stream into Huffman encoder
to do variable-bit+compression for us it might improve both blackbox
overhead and maximum logging time.

**Comment by ledvinap** [comment]
@digitalentity : Definitely worth a try ...

**Comment by hydra** [comment]
@etracer65 @ledvinap @martinbudden Transfer speed is limited by the MSP protocol.  IIRC when the EEPROM content is downloaded the configurator spends a lot of time doing nothing and so does the FC, mainly due to the MSP protocol implementation.  I think a lot of time could be saved by having the FC just stream out the EEPROM when requested rather than the way it's done now.

If you want to improve transfer speeds look there first before adding transfer compression.

**Comment by etracer65** [comment]
@hydra This is already been addressed.  Improvements in the packet size and increased baud rates are showing nearly a 7x increase in transfer speed.

Pull request: https://github.com/betaflight/betaflight/pull/1161

These optimizations came out of earlier discussions about using compression for blackbox downloads.

Issue: https://github.com/betaflight/betaflight/issues/1156

**Comment by hydra** [comment]
@etracer65 yes i thought i saw some comments about that, thanks for the links!

**Pull request closed without merging**